### PR TITLE
fix(factors): keep factor service alive after compute errors

### PR DESCRIPTION
**Pull request by stone16**
## Summary
- Return `None` from `AnchoringLagDivergence` before price validation when LLM/news inputs are absent, so book-style market-data signals with `yes_price=0.0` do not crash factor computation.
- Log and skip per-factor compute exceptions in `FactorService.compute_once()` instead of letting one factor kill the service task.
- Add regressions for missing H2 inputs on book signals and FactorService survival after a factor exception.

## Root Cause
`MarketDataSensor` can emit book signals without a usable last-trade price, producing `yes_price=0.0`. `AnchoringLagDivergence` validated `yes_price` before checking whether its own LLM/news inputs existed, raising `ValueError`. Since `FactorService` did not isolate factor compute exceptions, that single exception ended the background task while the rest of the runner kept reporting healthy.

## Validation
- `uv run pytest -q tests/unit/test_anchoring_lag_factor.py tests/unit/test_factor_service.py`
- `uv run pytest -q`
- `uv run mypy src/ tests/ --strict`
- `git diff --check`
- Restarted `pms-soak`; verified fresh `factor_values` rows continued past the original failure window.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved error handling in factor computation to continue processing remaining factors when individual computations fail, with detailed error logging.
  * Added signal stream exhaustion detection with appropriate warning notification.

* **Tests**
  * Added unit tests to verify robust error handling and edge case coverage.

### CODE REVIEW

**Comment by stone16** [comment]
Code Review Verdict: approve

Findings: none.

Reviewed `stone16/prediction-market-system#69` at head `fab7d04b52f1f28c09ba4a7d23dfcbd2142f1c8c`. The patch keeps `AnchoringLagDivergence` from validating placeholder book-signal prices when its H2 inputs are absent, and scopes the new FactorService exception isolation to per-factor compute failures while leaving persistence/service failures visible.

Verification run locally:
- `uv run pytest -q tests/unit/test_anchoring_lag_factor.py tests/unit/test_factor_service.py` -> 16 passed
- `uv run mypy src/ tests/ --strict` -> success, 403 files checked
- `git diff --check 4ebd40f803d5f1ae4e97fb8e3d266a1b6fe4b18c...HEAD` -> clean
- `uv run pytest -q` -> 1156 passed, 165 skipped

**Comment by stone16** [comment]
## Security & Performance Review — Dustin

**PR**: [stone16/prediction-market-system#69](https://github.com/stone16/prediction-market-system/pull/69) @ `fab7d04b52f1f28c09ba4a7d23dfcbd2142f1c8c`

### Security

- **No auth bypass, injection, secret leakage, side-channel, or concurrency safety issues.** The two changes are narrowly scoped to validation ordering and error isolation.
- The `except Exception` in `compute_once()` (`src/pms/factors/service.py:101`) correctly excludes `asyncio.CancelledError` (BaseException in 3.8+), so cooperative cancellation still propagates.
- Logged fields (`factor_id`, `market_id`, `timestamp`) at `src/pms/factors/service.py:102-106` are operational metadata only — no credential, token, or PII exposure.
- No new external dependencies — supply-chain footprint unchanged.

### Performance

- The new `try/except` at `src/pms/factors/service.py:99-108` has negligible overhead on the happy path (Python zero-cost try, no frame setup on entry).
- `logger.exception()` with `%s`-style formatting is lazy — no string allocation when the log level filters out ERROR.
- No new allocations in signal×factor hot loop. No blocking calls. No N+1 query patterns introduced.

### Non-blocking observations

1. **[log-flood risk]** `src/pms/factors/service.py:101-108` — A persistently-broken factor (e.g. TypeError from a schema mismatch) will log on every signal×factor iteration with no rate limiting. Under sustained error conditions with many signals, this can fill disk. Consider per-factor error deduplication or a circuit breaker that suppresses repeats after the first occurrence within a window.

2. **[missing-error-counter]** No metric for `factor_compute_errors_total{factor_id}`. Without it, a silently-failing factor requires active log-watching to notice. Good follow-up item — not blocking.

### Verdict

The reordering at `src/pms/factors/definitions/anchoring_lag_divergence.py:51-58` correctly pushes the (0,1)-open validation past the null-input guard. The `try/except` isolation at `src/pms/factors/service.py:99-108` correctly prevents one factor's exception from killing the service background task. Both changes are covered by targeted regression tests. No security or performance regressions.

**approve**

File: src/pms/factors/definitions/anchoring_lag_divergence.py [hunk 1]
```diff
@@ -48,14 +48,14 @@ def compute(
...
         del outer_ring
 
-        yes_price = _require_open_probability(signal.yes_price, "yes_price")
         if (
             "llm_posterior" not in signal.external_signal
             or "news_timestamp" not in signal.external_signal
             or signal.external_signal["llm_posterior"] is None
             or signal.external_signal["news_timestamp"] is None
         ):
             return None
```
[the remark this answers is not in the document]
Comment (stone16) [reply src/pms/factors/definitions/anchoring_lag_divergence.py:57]: Fixed in `9d4efc7`.

You were right: suppressing only `AnchoringLagDivergence` still allowed other `yes_price` consumers to run on the `MarketDataSensor` sentinel `yes_price=0.0`. I moved this to a FactorService-level required-input gate: factors declaring `required_inputs` contains `yes_price` are skipped unless the signal has a real open probability, while orderbook-only factors still compute.

Added `test_factor_service_skips_yes_price_factors_for_sentinel_book_signals` to cover this exact path.

Validation:
- `uv run pytest -q tests/unit/test_anchoring_lag_factor.py tests/unit/test_factor_service.py` -> 17 passed
- `uv run pytest -q` -> 1157 passed, 165 skipped
- `uv run mypy src/ tests/ --strict` -> success

**Comment by stone16** [comment]
## Security & Performance Review (Dustin)

### Security
- **Auth boundaries**: No changes — N/A.
- **Dependencies**: No new dependencies — N/A.
- **Secrets handling**: No changes — N/A.
- **Concurrency safety**: No regressions. The try/except isolation in `compute_once` operates on per-iteration locals; no new shared-state races.
- **Input validation**: Correct. `_has_open_yes_price` (`src/pms/factors/service.py:214-216`) uses `0.0 < yes_price < 1.0 and yes_price == yes_price` — the latter is the standard Python NaN guard (NaN != NaN). The root-cause fix in `AnchoringLagDivergence.compute()` (`src/pms/factors/definitions/anchoring_lag_divergence.py:51-57`) correctly moves the early-return *before* `_require_open_probability`, so book signals with `yes_price=0.0` and no LLM/news inputs return `None` instead of raising `ValueError`.

### Minor finding (non-blocking)
**`[guard-outside-try]`** `src/pms/factors/service.py:99` — The `_has_open_yes_price(signal)` guard runs outside the `try/except` at line 103. If `yes_price` were somehow non-numeric at runtime (despite the `float` type annotation on the frozen `MarketSignal` dataclass), the `<` comparison would raise `TypeError` and crash `compute_once`. The frozen-dataclass type system makes this extremely unlikely; defense-in-depth would move the guard inside the try/except. Not blocking.

### Performance
- The `_requires_open_yes_price` / `_has_open_yes_price` gates (`src/pms/factors/service.py:210-216`) are O(1) in-memory checks that skip factor computation for signals without a usable price — a strict improvement over letting `compute()` raise and catching the exception.
- No new database queries, no N+1 regressions, no new allocations in hot paths.
- The `try/except` in `compute_once` adds negligible overhead on the happy path (Python's try/except is zero-cost when no exception is raised).

### Test coverage
The three new tests are well-targeted:
- `test_anchoring_lag_factor_ignores_book_signal_without_llm_news_inputs` — verifies the root fix (yes_price=0.0 + no external_signal → None, no crash)
- `test_factor_service_compute_once_continues_after_factor_error` — verifies `compute_once` survives an `ExplodingFactor` and still processes `PersistedFactor`
- `test_factor_service_skips_yes_price_factors_for_sentinel_book_signals` — verifies the yes_price gate skips `PersistedFactor` but runs `OrderbookOnlyFactor` for 0.0-price signals

### Verdict: **approve**

Changes are minimal, correct, and well-tested. No security or performance regressions.

**Comment by stone16** [comment]
Code Review Verdict: approve

Findings: none.

Reviewed `stone16/prediction-market-system#69` at current head `9d4efc7f9647bf211a2bf051c69ef7457a5c7a3a`. The follow-up commit moves sentinel `yes_price=0.0` handling to the FactorService input gate for all factors declaring `yes_price`, while keeping orderbook-only factors eligible and preserving persistence/service failures outside the per-factor compute exception isolation.

Verification run locally:
- `uv run pytest -q tests/unit/test_anchoring_lag_factor.py tests/unit/test_factor_service.py` -> 17 passed
- `uv run mypy src/ tests/ --strict` -> success, 403 files checked
- `git diff --check origin/main...HEAD` -> clean
- `uv run pytest -q` -> 1157 passed, 165 skipped
